### DISCUSSION TITLE
Where could I get reference book about FE shape functions?

### DISCUSSION THREAD

**Discussion by Richard-happy** [libMesh/libmesh#3805] [Q&A]
As the title says, I found there were lots of types of FE shape functions in libmesh, such as monomial, hierarchic, szabab and so on. I'm confused about which type of  FE shape functions I should choose when solving different problems. Could you please introduce me an outstanding reference book?
Looking forward to your reply.

**Comment by roystgnr** [comment]
I recently ran across https://defelement.com/, which is rather interesting.  It's not a perfectly libMesh-compatible reference, though - for instance, it has the Bogner-Fox-Schmidt element, but we just call that "Hermite" since it's a tensor product of 1-D Hermites.  They also don't mention the interesting limitations of BFS (you can't match up that mixed second derivative properly on arbitrary quadrilateral meshes, you really need something rectilinear), and they don't have any extensions to higher polynomial degree like we do, and I don't even see the 3D version of those elements there.

The choice of shape functions to solve different problems can be really problem dependent if you don't have a stabilized formulation.  You're probably best off looking for papers talking about the formulation you want to solve; you should find discussion of any stability concerns there, whereas if you just look for a reference to a FE type you're not going to find discussion of the thousand different problems it could be (or couldn't be) used for.

If you have an H1 problem with no stability issues then I'd just start off with MONOMIAL for any discontinuous variables, and HIERARCHIC for any C0 variables you want to do p>2 with; if you're sure you don't need high p then LAGRANGE might be simpler to work with.

My favorite reference for weird FEM issues (as well as standard issues like LBB stability, IIRC) was Becker, Carey, and Oden, but I think the series is (mostly? maybe you can still get volume 1?) long out of print now, and I haven't found anything else that will happily explain, for example, why the standard "Nitsche lift" trick doesn't actually promise the convergence you'd naively expect from a p=2 C1 element on an H2 problem.

**Reply by mscroggs** [reply]
> or instance, it has the Bogner-Fox-Schmidt element, but we just call that "Hermite" since it's a tensor product of 1-D Hermites. They also don't mention the interesting limitations of BFS (you can't match up that mixed second derivative properly on arbitrary quadrilateral meshes, you really need something rectilinear), and they don't have any extensions to higher polynomial degree like we do, and I don't even see the 3D version of those elements there.

I've opened [an issue on the DefElement repo](https://github.com/DefElement/DefElement/issues/391) for these points. Feel free to add to it if you have references, etc to add
